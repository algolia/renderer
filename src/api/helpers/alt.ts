import Altheia from 'altheia-async-data-validator';

export const alt = Altheia.instance();
alt.lang('protocol_not_allowed', () => 'Only HTTP protocol is allowed');

export function getDefaultParams(): Record<any, any> {
  return {
    url: alt
      .internet()
      .url()
      .custom('protocol_not_allowed', (val) => {
        return ['http:', 'https:'].includes(new URL(val).protocol);
      })
      .required(),
    ua: alt.string().required(),
    waitTime: alt.object().schema(
      alt({
        min: alt.number().cast().min(1000).max(19000),
        max: alt.number().cast().min(2000).max(20000),
      })
    ),
  };
}
