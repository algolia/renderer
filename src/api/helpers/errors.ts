import type express from 'express';

interface AnyParams {
  res: express.Response;
  status: number;
  message: string;
  details?: any;
}

function any({ res, status, message, details }: AnyParams): void {
  res.status(status).json({ error: true, message, details });
}

interface BadRequestParams {
  res: express.Response;
  message?: string;
  details?: any;
}

export function badRequest({
  res,
  details,
  message = 'Bad Request',
}: BadRequestParams): void {
  return any({
    res,
    status: 400,
    message,
    details,
  });
}
