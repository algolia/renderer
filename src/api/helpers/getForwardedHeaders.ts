import type express from 'express';

import { HEADERS_TO_FORWARD } from 'api/constants';

export function getForwardedHeadersFromRequest(
  req: express.Request<any, any, any, any>
): Record<string, string> {
  const headersToForward = HEADERS_TO_FORWARD.reduce((partial, headerName) => {
    const name = headerName.toLowerCase();
    if (req.headers[name]) {
      return { ...partial, [name]: req.headers[name] };
    }
    return partial;
  }, {});

  return headersToForward;
}
