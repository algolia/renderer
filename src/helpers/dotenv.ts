export default () => {
  if (process.env.NODE_ENV === 'development') {
    require('dotenv-safe').config({ allowEmptyValues: true });
  }
};
