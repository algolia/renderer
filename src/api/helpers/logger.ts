import { log as mainLog } from 'helpers/logger';

export const log = mainLog.child({ svc: 'api ' });
