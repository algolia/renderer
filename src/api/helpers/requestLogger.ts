import type express from 'express';

import { log } from './logger';

export function requestLogger(
  req: express.Request,
  res: express.Response,
  next: express.NextFunction
): void {
  if (['/ready', '/healthy'].includes(req.url)) {
    next();
    return;
  }

  log.info('New request', { method: req.method, path: req.url });
  next();
}
