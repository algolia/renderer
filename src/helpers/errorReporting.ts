import * as Sentry from '@sentry/node';

import { log } from './logger';

Sentry.init({
  dsn: process.env.SENTRY_DSN,
  release: process.env.npm_package_version,
  environment: process.env.CLUSTER_NAME || process.env.NODE_ENV,
  serverName: 'renderscript',
  ignoreErrors: [],
  maxBreadcrumbs: 10,
});

export function report(err: Error, extra: any = {}): void {
  if (!process.env.SENTRY_DSN) {
    log.error(err.message, err.stack, extra);
    return;
  }

  log.error(err.message);
  Sentry.withScope((scope) => {
    scope.setExtras(extra);
    Sentry.captureException(err);
  });
}

export async function drain(): Promise<boolean> {
  const client = Sentry.getCurrentHub().getClient();
  if (client) {
    return await client.close(2000);
  }

  return true;
}
