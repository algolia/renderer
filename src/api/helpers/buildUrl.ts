import { report } from 'helpers/errorReporting';

const DOCKER_LOCALHOST = 'host.docker.internal';

const USE_DOCKER_LOCALHOST = process.env.USE_DOCKER_LOCALHOST === 'true';

export function replaceHost(url: URL, from: string, to: string): URL {
  const fromRegex = new RegExp(`^${from}(:|$)`);
  const host = url.host || '';
  // eslint-disable-next-line no-param-reassign
  url.host = host.replace(fromRegex, `${to}$1`);
  return url;
}

export function revertUrl(href: string | null): URL | null {
  if (!href) {
    return null;
  }

  try {
    const url = new URL(href);
    if (!USE_DOCKER_LOCALHOST) {
      return url;
    }
    return replaceHost(url, DOCKER_LOCALHOST, 'localhost');
  } catch (err) {
    report(new Error('invalid revertUrl'), { href });
    return null;
  }
}

export function buildUrl(href: string): URL {
  const url = new URL(href);
  if (!USE_DOCKER_LOCALHOST) {
    return url;
  }
  return replaceHost(url, 'localhost', DOCKER_LOCALHOST);
}
