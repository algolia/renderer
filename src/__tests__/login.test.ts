import type { Cookie } from 'playwright-chromium';

import type { PostLoginSuccess } from 'api/@types/postLogin';

import { sendLoginRequest } from './helpers';

jest.setTimeout(45000);

describe('login', () => {
  it('should error when no username', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login',
      username: '',
      password: 'password',
    });

    expect(res.statusCode).toBe(400);

    expect(JSON.parse(body)).toEqual({
      details: [
        {
          label: 'username',
          message: 'username is required',
          type: 'required',
        },
      ],
      error: true,
      message: 'Bad Request',
    });
  });

  it('should error when no password', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login',
      username: 'admin',
      password: '',
    });

    expect(res.statusCode).toBe(400);

    expect(JSON.parse(body)).toEqual({
      details: [
        {
          label: 'password',
          message: 'password is required',
          type: 'required',
        },
      ],
      error: true,
      message: 'Bad Request',
    });
  });

  it('should error double text input', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/test-website/login-double-input.html',
      username: 'admin',
      password: 'paswword',
    });

    expect(res.statusCode).toBe(200);
    const parsed: PostLoginSuccess = JSON.parse(body);
    expect(parsed.error).toBe('too_many_fields');
    expect(parsed.rawError?.message).toBe(
      'Too many input found for "input[type=text], input[type=email]", found "2"'
    );
  });

  it('should error double password', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/test-website/login-double-password.html',
      username: 'admin',
      password: 'paswword',
    });

    expect(res.statusCode).toBe(200);
    const parsed: PostLoginSuccess = JSON.parse(body);
    expect(parsed.error).toBe('too_many_fields');
    expect(parsed.rawError?.message).toBe(
      'Too many input found for "input[type=password]:not([aria-hidden="true"])", found "2"'
    );
  });

  it('should works with correct credentials', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login',
      username: 'admin',
      password: 'password',
    });

    expect(res.statusCode).toBe(200);

    const parsed: PostLoginSuccess = JSON.parse(body);
    expect(
      parsed.cookies.find((cookie) => cookie.name === 'sessionToken')
    ).toMatchSnapshot();
    // Check that we actually went through the form
    expect(
      parsed.cookies.find((cookie) => cookie.name === '_csrf')
    ).toBeDefined();
  });

  it('should works even with a 2-steps login', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login/step1',
      username: 'admin',
      password: 'password',
    });

    expect(res.statusCode).toBe(200);

    const cookies: Cookie[] = JSON.parse(body).cookies;
    expect(
      cookies.find((cookie) => cookie.name === 'sessionToken')
    ).toMatchSnapshot();
    // Check that we actually went through the form
    expect(cookies.find((cookie) => cookie.name === '_csrf')).toBeDefined();
  });

  it('should works with a 2-steps JS login', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login/2steps',
      username: 'admin',
      password: 'password',
    });

    expect(res.statusCode).toBe(200);

    const cookies: Cookie[] = JSON.parse(body).cookies;
    expect(
      cookies.find((cookie) => cookie.name === 'sessionToken')
    ).toMatchSnapshot();
    // Check that we actually went through the form
    expect(cookies.find((cookie) => cookie.name === '_csrf')).toBeDefined();
  });

  it('should works but not get a session token with bad credentials', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login',
      username: 'admin',
      password: 'admin',
    });

    expect(res.statusCode).toBe(200);

    const parsed: PostLoginSuccess = JSON.parse(body);
    expect(parsed.cookies).toHaveLength(1);
    expect(
      parsed.cookies.find((cookie) => cookie.name === 'sessionToken')
    ).toBeUndefined();
    // Check that we actually went through the form
    expect(
      parsed.cookies.find((cookie) => cookie.name === '_csrf')
    ).toBeDefined();
  });
});

describe('JavaScript redirect', () => {
  it('should fail to renderHTML because of the JS redirect', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login?redirect=true',
      username: 'admin',
      password: 'password',
      renderHTML: true,
      waitTime: {
        min: 1000,
      },
    });

    expect(res.statusCode).toBe(200);
    expect(body).toBe(
      '<!DOCTYPE html><html lang="en"><head></head><body>OK(/test)</body></html>'
    );
  });

  it('should not try to render the body if renderHTML was not requested', async () => {
    const { res, body } = await sendLoginRequest({
      url: 'http://localhost:3000/secure/login?redirect=true',
      username: 'admin',
      password: 'password',
      waitTime: {
        min: 1000,
      },
    });

    // Since we didn't try to render, it returns the current cookies, even if there is an ongoing JS redirection
    expect(res.statusCode).toBe(200);

    const parsed: PostLoginSuccess = JSON.parse(body);
    expect(parsed.body).toBe(
      '<!DOCTYPE html><html lang="en"><head></head><body>OK(/test)</body></html>'
    );
    expect(parsed.statusCode).toBe(200);
    expect(parsed.metrics.timings.total).toBeGreaterThan(1000);
    expect(parsed.resolvedUrl).toBe('http://localhost:3000/secure/test');
    expect(
      parsed.cookies.find((cookie) => cookie.name === 'sessionToken')
    ).toMatchSnapshot();
  });
});
