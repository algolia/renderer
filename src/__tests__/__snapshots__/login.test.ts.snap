// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JavaScript redirect should not try to render the body if renderHTML was not requested 1`] = `
Object {
  "domain": "localhost",
  "expires": -1,
  "httpOnly": false,
  "name": "sessionToken",
  "path": "/secure",
  "sameSite": "Strict",
  "secure": false,
  "value": "53cu23_535510n",
}
`;

exports[`login should works even with a 2-steps login 1`] = `
Object {
  "domain": "localhost",
  "expires": -1,
  "httpOnly": false,
  "name": "sessionToken",
  "path": "/secure",
  "sameSite": "Strict",
  "secure": false,
  "value": "53cu23_535510n",
}
`;

exports[`login should works with a 2-steps JS login 1`] = `
Object {
  "domain": "localhost",
  "expires": -1,
  "httpOnly": false,
  "name": "sessionToken",
  "path": "/secure",
  "sameSite": "Strict",
  "secure": false,
  "value": "53cu23_535510n",
}
`;

exports[`login should works with correct credentials 1`] = `
Object {
  "domain": "localhost",
  "expires": -1,
  "httpOnly": false,
  "name": "sessionToken",
  "path": "/secure",
  "sameSite": "Strict",
  "secure": false,
  "value": "53cu23_535510n",
}
`;
